plugins/node/opentelemetry-instrumentation-cassandra/src/instrumentation.ts

```diff
@@ -0,0 +1,366 @@
+              return original.apply(this, args);
+            });
+          },
+          error => {
+            if (error) {
+              failSpan(span, error);
+            }
+          }
+        );
+
+        const wrappedPromise = wrapPromise(span, execPromise);
+
+        return context.bind(execContext, wrappedPromise);
+      };
+    };
+  }
+
+  private _getPatchedSendOnConnection() {
+    return (original: (...args: unknown[]) => unknown) => {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      return function patchedSendOnConnection(this: any, ...args: unknown[]) {
+        const span = trace.getSpan(context.active());
+        const conn = this['_connection'];
+
+        if (span !== undefined && conn !== undefined) {
+          const port = parseInt(conn.port, 10);
+
+          span.setAttribute(SemanticAttributes.NET_PEER_NAME, conn.address);
+
+          if (!isNaN(port)) {
+            span.setAttribute(SemanticAttributes.NET_PEER_PORT, port);
+          }
+        }
+
+        return original.apply(this, args);
+      };
+    };
+  }
+
+  private _getPatchedBatch() {
+    return (original: (...args: unknown[]) => unknown) => {
+      const plugin = this;
+      return function patchedBatch(
+        this: CassandraDriver.Client,
+        ...args: unknown[]
+      ) {
+        const queries = Array.isArray(args[0]) ? args[0] : [];
+        const combined = truncateQuery(
+          combineQueries(queries),
+          plugin.getMaxQueryLength()
+        );
+        const span = startSpan(plugin.tracer, 'batch', combined, this);
+
+        const batchContext = trace.setSpan(context.active(), span);
+
+        if (typeof args[args.length - 1] === 'function') {
+          const originalCallback = args[
+            args.length - 1
+          ] as CassandraDriver.ValueCallback<CassandraDriver.types.ResultSet>;
+
+          const patchedCallback = function (
+            this: unknown,
+            ...cbArgs: Parameters<typeof originalCallback>
+          ) {
+            const error = cbArgs[0];
+
+            if (error) {
+              span.setStatus({
+                code: SpanStatusCode.ERROR,
+                message: error.message,
+              });
+              span.recordException(error);
+            }
+
+            span.end();
+
+            return originalCallback.apply(this, cbArgs);
+          };
+
+          args[args.length - 1] = patchedCallback;
+
+          return context.with(batchContext, () => {
+            return original.apply(this, args);
+          });
+        }
+
+        const batchPromise = safeExecuteInTheMiddle(
+          () => {
+            return original.apply(
+              this,
+              args
+            ) as Promise<CassandraDriver.types.ResultSet>;
+          },
+          error => {
+            if (error) {
+              failSpan(span, error);
+            }
+          }
+        );
+
+        const wrappedPromise = wrapPromise(span, batchPromise);
```

Comment:
since you are wrapping the promise, do we need `safeExecuteInTheMiddle` ? the promise can't throw if you catch it right ?

Reply:
The call itself might throw (e.g. on invalid input)